tests/gocase/unit/type/list/list_test.go

```diff
@@ -39,6 +39,118 @@ var largeValue = map[string]string{
+
+	t.Run("LPUSH, RPUSH, LLENGTH, LINDEX, LPOP - regular list", func(t *testing.T) {
+		myList1 := "mylist1"
+		myList2 := "mylist2"
+
+		linkedListValue := largeValue["linkedList"]
+
+		// first lpush then rpush
+		require.EqualValues(t, 1, rdb.LPush(ctx, myList1, linkedListValue).Val())
+		require.EqualValues(t, 2, rdb.RPush(ctx, myList1, "b").Val())
+		require.EqualValues(t, 3, rdb.RPush(ctx, myList1, "c").Val())
+		require.EqualValues(t, 3, rdb.LLen(ctx, myList1).Val())
+		require.EqualValues(t, linkedListValue, rdb.LIndex(ctx, myList1, 0).Val())
+		require.EqualValues(t, "b", rdb.LIndex(ctx, myList1, 1).Val())
+		require.EqualValues(t, "c", rdb.LIndex(ctx, myList1, 2).Val())
+		require.EqualValues(t, "", rdb.LIndex(ctx, myList1, 3).Val())
+		require.EqualValues(t, "c", rdb.RPop(ctx, myList1).Val())
+		require.EqualValues(t, linkedListValue, rdb.LPop(ctx, myList1).Val())
+
+		// first rpush then lpush
+		require.EqualValues(t, 1, rdb.RPush(ctx, myList2, linkedListValue).Val())
+		require.EqualValues(t, 2, rdb.LPush(ctx, myList2, "b").Val())
+		require.EqualValues(t, 3, rdb.LPush(ctx, myList2, "c").Val())
+		require.EqualValues(t, 3, rdb.LLen(ctx, myList2).Val())
+
+		require.EqualValues(t, "c", rdb.LIndex(ctx, myList2, 0).Val())
+		require.EqualValues(t, "b", rdb.LIndex(ctx, myList2, 1).Val())
+		require.EqualValues(t, linkedListValue, rdb.LIndex(ctx, myList2, 2).Val())
+		require.EqualValues(t, "", rdb.LIndex(ctx, myList2, 3).Val())
+		require.EqualValues(t, linkedListValue, rdb.RPop(ctx, myList2).Val())
+		require.EqualValues(t, "c", rdb.LPop(ctx, myList2).Val())
+	})
+
+	t.Run("R&LPOP against empty list", func(t *testing.T) {
+		require.Error(t, rdb.LPop(ctx, "non-existing-list").Err())
+	})
+
+	t.Run("Variadic RPUSH&LPUSH", func(t *testing.T) {
+		myList := "mylist"
+		require.NoError(t, rdb.Del(ctx, myList).Err())
+		require.EqualValues(t, 4, rdb.LPush(ctx, myList, "a", "b", "c", "d").Val())
+		require.EqualValues(t, 8, rdb.RPush(ctx, myList, "0", "1", "2", "3").Val())
+		require.Equal(t, []string{"d", "c", "b", "a", "0", "1", "2", "3"}, rdb.LRange(ctx, myList, 0, -1).Val())
+	})
+
+	t.Run("DEL a list", func(t *testing.T) {
+		myList2 := "mylist2"
+		require.EqualValues(t, 1, rdb.Del(ctx, myList2).Val())
+		require.EqualValues(t, 0, rdb.Exists(ctx, myList2).Val())
+		require.EqualValues(t, 0, rdb.LLen(ctx, myList2).Val())
+	})
+
+	createList := func(key string, entries []string) {
+		require.NoError(t, rdb.Del(ctx, key).Err())
+		for _, v := range entries {
+			require.NoError(t, rdb.RPush(ctx, key, v).Err())
+		}
+	}
+
+	for typ, value := range largeValue {
+		t.Run(fmt.Sprintf("BLPOP, BRPOP: single existing list - %s", typ), func(t *testing.T) {
+			c := srv.NewTCPClient()
+			defer func() { require.NoError(t, c.Close()) }()
+			createList("blist", []string{"a", "b", value, "c", "d"})
+			res := rdb.LRange(ctx, "blist", 0, -1).Val()
+			fmt.Println(res)
+			require.NoError(t, c.Write("*3\r\n$5\r\nBLPOP\r\n$5\r\nblist\r\n$1\r\n1\r\n"))
```

Comment:
I used raw protocol here.

Reply:
@tisonkun Hi, I found out that in `list.tcl`, some cases use `set rd [redis_deferring_client]` as a deferring client.

I wonder how to use a "deferring client" in Go.

---

The study I have done:
- I have tried `NewTCPClient`, and although it has similar APIs (such as `ReadLine`), but it turns out to be that I have to write raw Redis protocol to communicate with the server. Actually it works, but it seems to be a little bit awkward... Any way, I have written a little piece of code at [here](https://github.com/apache/incubator-kvrocks/pull/866/files#diff-6df089ad841b180ccd3a65619ee756b051f26ed20ea3cc984feb985c41c9da02). Please comment on this if I did anything wrong :)
- I also have learned how you dealt with the deferring client in #854 and #810, and it seems to be we still need to write raw protocol in code.